trinity/bootstrap.py

```diff
@@ -0,0 +1,282 @@
+        args.stderr_log_level or (args.log_levels and args.log_levels.get(None))
+    )
+
+    if args.log_levels:
+        setup_log_levels(args.log_levels)
+
+    try:
+        trinity_config = TrinityConfig.from_parser_args(args)
+    except AmbigiousFileSystem:
+        parser.error(TRINITY_AMBIGIOUS_FILESYSTEM_INFO)
+
+    if not is_data_dir_initialized(trinity_config):
+        # TODO: this will only work as is for chains with known genesis
+        # parameters.  Need to flesh out how genesis parameters for custom
+        # chains are defined and passed around.
+        try:
+            initialize_data_dir(trinity_config)
+        except AmbigiousFileSystem:
+            parser.error(TRINITY_AMBIGIOUS_FILESYSTEM_INFO)
+        except MissingPath as e:
+            parser.error(
+                "\n"
+                f"It appears that {e.path} does not exist. "
+                "Trinity does not attempt to create directories outside of its root path. "
+                "Either manually create the path or ensure you are using a data directory "
+                "inside the XDG_TRINITY_ROOT path"
+            )
+
+    file_logger, log_queue, listener = setup_trinity_file_and_queue_logging(
+        stderr_logger,
+        formatter,
+        handler_stream,
+        trinity_config.logfile_path,
+        args.file_log_level,
+    )
+
+    display_launch_logs(trinity_config)
+
+    # compute the minimum configured log level across all configured loggers.
+    min_configured_log_level = min(
+        stderr_logger.level,
+        file_logger.level,
+        *(args.log_levels or {}).values()
+    )
+
+    extra_kwargs = {
+        'log_queue': log_queue,
+        'log_level': min_configured_log_level,
+        'profile': args.profile,
+    }
+
+    # Plugins can provide a subcommand with a `func` which does then control
+    # the entire process from here.
+    if hasattr(args, 'func'):
+        args.func(args, trinity_config)
+    else:
+        trinity_boot(
+            args,
+            trinity_config,
+            extra_kwargs,
+            plugin_manager,
+            listener,
+            event_bus,
+            main_endpoint,
+            stderr_logger,
+        )
+
+
+def setup_plugins(scope: BaseManagerProcessScope) -> PluginManager:
+    plugin_manager = PluginManager(scope)
+    # TODO: most plugins should check if they are in eth1 / eth2 context
+    # and only start when appropriate
+    plugin_manager.register(get_all_plugins())
+
+    return plugin_manager
+
+
+def display_launch_logs(trinity_config: TrinityConfig) -> None:
+    logger = logging.getLogger('trinity')
+    logger.info(TRINITY_HEADER)
+    logger.info("Started main process (pid=%d)", os.getpid())
+    logger.info(construct_trinity_client_identifier())
+    logger.info("Trinity DEBUG log file is created at %s", str(trinity_config.logfile_path))
+
+
+@setup_cprofiler('run_database_process')
```

Comment:
side-note.

I think this is junk and we could drop it.  The way it works I don't believe it provides us with very useful profiling data (but my memory may be wrong here).

Reply:
I just tried running `trinity --profile` and it produces two profile dumps and after installing `cprofilev` I can simply run `cprofilev -f <name-of-dump-file` to start inspecting.

![image](https://user-images.githubusercontent.com/521109/49506315-14015700-f87e-11e8-98c8-f0a84d17ea83.png)

So, I actually think this is very valuable and should be documented better. Or is there something else that makes you believe the data itself isn't useful?

Reply:
I created a separate issue for this to not block this PR. https://github.com/ethereum/py-evm/issues/1548

Reply:
I seem to recall that one deficiency is that it doesn't cover things which run in our process executors.  Otherwise :+1: that it does what we want.